File: android/app/src/test/java/dev/ironkeep/app/vault/model/VaultMutationsTest.kt
package dev.ironkeep.app.vault.model

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant

class VaultMutationsTest {
    @Test
    fun categoriesAndTagsDoNotLeaveDanglingItemReferences() {
        val login = VaultMutations.addLogin(
            VaultPayload.empty("Test", "device-a"),
            LoginFields("Example", "person@example.com", "secret", emptyList(), emptyList()),
            "device-a",
            itemId = "login-a",
        )
        val withCategory = VaultMutations.addCategory(login, "Finance", "device-a")
        val category = withCategory.categories.first { it.name == "Finance" }
        val withTag = VaultMutations.addTag(withCategory, "Important", "device-a")
        val tag = withTag.tags.first { it.name == "Important" }
        val organized = VaultMutations.setItemOrganization(withTag, "login-a", category.id, listOf(tag.id), "device-a")
        assertEquals(category.id, organized.items.single().categoryId)
        assertEquals(listOf(tag.id), organized.items.single().tagIds)

        val cleaned = VaultMutations.deleteTag(VaultMutations.deleteCategory(organized, category.id, "device-a"), tag.id, "device-a")
        assertNull(cleaned.items.single().categoryId)
        assertTrue(cleaned.items.single().tagIds.isEmpty())
    }

    private val fields = LoginFields(
        title = "Example",
        username = "person@example.com",
        password = "secret",
        uris = listOf("https://example.com/login"),
        androidPackageNames = listOf("com.example.app"),
    )

    @Test
    fun addEditFavoriteAndDeleteUpdateMetadataAndCreateTombstone() {
        val empty = VaultPayload.empty("Test", "device-a").copy(revision = 1, updatedAt = "2026-01-01T00:00:00Z")
        val added = VaultMutations.addLogin(empty, fields, "device-a", Instant.parse("2026-01-02T00:00:00Z"), "login-one")
        assertEquals(2, added.revision)
        assertEquals(1, (added.items.single() as LoginItem).revision)

        val edited = VaultMutations.editLogin(added, "login-one", fields.copy(title = "Updated"), "device-b", Instant.parse("2026-01-03T00:00:00Z"))
        assertEquals(3, edited.revision)
        assertEquals("device-b", edited.writerDeviceId)
        assertEquals(2, (edited.items.single() as LoginItem).revision)

        val favorite = VaultMutations.toggleFavorite(edited, "login-one", "device-b", Instant.parse("2026-01-04T00:00:00Z"))
        assertTrue((favorite.items.single() as LoginItem).favorite)

        val deleted = VaultMutations.deleteLogin(favorite, "login-one", "device-b", Instant.parse("2026-01-05T00:00:00Z"))
        assertTrue(deleted.items.isEmpty())
        assertEquals(Tombstone("login-one", "2026-01-05T00:00:00Z", 5, "device-b"), deleted.tombstones.single())
    }

    @Test
    fun duplicateDetectionRequiresSameIdentityAndExactOriginOrPackage() {
        val vault = VaultMutations.addLogin(VaultPayload.empty("Test", "device-a"), fields, "device-a", itemId = "login-one")
        assertEquals(1, VaultMutations.likelyDuplicates(vault, fields.copy(uris = listOf("https://example.com/other"), androidPackageNames = emptyList())).size)
        assertFalse(VaultMutations.likelyDuplicates(vault, fields.copy(title = "Other", username = "other@example.com", uris = listOf("https://other.example"), androidPackageNames = emptyList())).isNotEmpty())
    }

    @Test
    fun securitySettingsUpdateVaultMetadataOnce() {
        val empty = VaultPayload.empty("Test", "device-a").copy(revision = 1, updatedAt = "2026-01-01T00:00:00Z")
        val updated = VaultMutations.updateSecuritySettings(empty, 15, 60, "device-b", Instant.parse("2026-01-02T00:00:00Z"))
        assertEquals(2, updated.revision)
        assertEquals("2026-01-02T00:00:00Z", updated.updatedAt)
        assertEquals("device-b", updated.writerDeviceId)
        assertEquals(15, updated.settings.autoLockMinutes)
        assertEquals(60, updated.settings.clearClipboardSeconds)
    }

    @Test
    fun secureNoteCrudPreservesMetadataAndCreatesTombstone() {
        val empty = VaultPayload.empty("Test", "device-a").copy(revision = 1, updatedAt = "2026-01-01T00:00:00Z")
        val added = VaultMutations.addSecureNote(empty, SecureNoteFields("Recovery", "Offline codes"), "device-a", Instant.parse("2026-01-02T00:00:00Z"), "note-one")
        assertEquals(SecureNoteItem::class, added.items.single()::class)
        assertEquals(1, added.items.single().revision)

        val edited = VaultMutations.editSecureNote(added, "note-one", SecureNoteFields("Recovery codes", "Updated codes"), "device-b", Instant.parse("2026-01-03T00:00:00Z"))
        val note = edited.items.single() as SecureNoteItem
        assertEquals("Recovery codes", note.title)
        assertEquals("Updated codes", note.body)
        assertEquals(2, note.revision)
        assertEquals("2026-01-02T00:00:00Z", note.createdAt)

        val favorite = VaultMutations.toggleSecureNoteFavorite(edited, "note-one", "device-b", Instant.parse("2026-01-04T00:00:00Z"))
        assertTrue((favorite.items.single() as SecureNoteItem).favorite)

        val deleted = VaultMutations.deleteSecureNote(favorite, "note-one", "device-b", Instant.parse("2026-01-05T00:00:00Z"))
        assertTrue(deleted.items.isEmpty())
        assertEquals(Tombstone("note-one", "2026-01-05T00:00:00Z", 5, "device-b"), deleted.tombstones.single())
    }

    @Test
    fun secureNoteValidationAndDuplicateDetectionFailClosed() {
        val empty = VaultPayload.empty("Test", "device-a")
        assertTrue(runCatching { VaultMutations.addSecureNote(empty, SecureNoteFields("", "body"), "device-a") }.isFailure)
        assertTrue(runCatching { VaultMutations.addSecureNote(empty, SecureNoteFields("Title", " "), "device-a") }.isFailure)
        val added = VaultMutations.addSecureNote(empty, SecureNoteFields("Recovery", "codes"), "device-a", itemId = "note-one")
        assertEquals(1, VaultMutations.likelySecureNoteDuplicates(added, SecureNoteFields(" recovery ", "different")).size)
        assertTrue(VaultMutations.likelySecureNoteDuplicates(added, SecureNoteFields("Other", "different")).isEmpty())
    }

    @Test
    fun creditCardCrudNormalizesSecretsAndCreatesTombstone() {
        val card = CreditCardFields("Travel card", "A Person", "4111 1111 1111 1111", 12, 2030, "123", "9876", "Use abroad")
        val empty = VaultPayload.empty("Test", "device-a").copy(revision = 1, updatedAt = "2026-01-01T00:00:00Z")
        val added = VaultMutations.addCreditCard(empty, card, "device-a", Instant.parse("2026-01-02T00:00:00Z"), "card-one")
        assertEquals("4111111111111111", (added.items.single() as CreditCardItem).number)
        val edited = VaultMutations.editCreditCard(added, "card-one", card.copy(title = "Primary card", notes = "Updated"), "device-b", Instant.parse("2026-01-03T00:00:00Z"))
        assertEquals(2, (edited.items.single() as CreditCardItem).revision)
        val favorite = VaultMutations.toggleCreditCardFavorite(edited, "card-one", "device-b", Instant.parse("2026-01-04T00:00:00Z"))
        assertTrue((favorite.items.single() as CreditCardItem).favorite)
        val deleted = VaultMutations.deleteCreditCard(favorite, "card-one", "device-b", Instant.parse("2026-01-05T00:00:00Z"))
        assertTrue(deleted.items.isEmpty())
        assertEquals(Tombstone("card-one", "2026-01-05T00:00:00Z", 5, "device-b"), deleted.tombstones.single())
    }

    @Test
    fun creditCardValidationAndDuplicateDetectionFailClosed() {
        val card = CreditCardFields("Travel card", "A Person", "4111111111111111", 12, 2030, "123", null, "")
        val empty = VaultPayload.empty("Test", "device-a")
        assertTrue(runCatching { VaultMutations.addCreditCard(empty, card.copy(number = "123"), "device-a") }.isFailure)
        assertTrue(runCatching { VaultMutations.addCreditCard(empty, card.copy(expiryMonth = 13), "device-a") }.isFailure)
        assertTrue(runCatching { VaultMutations.addCreditCard(empty, card.copy(verificationCode = "x"), "device-a") }.isFailure)
        val added = VaultMutations.addCreditCard(empty, card, "device-a", itemId = "card-one")
        assertEquals(1, VaultMutations.likelyCreditCardDuplicates(added, card.copy(number = "4111-1111-1111-1111")).size)
    }

    @Test
    fun identityCrudPreservesFieldsAndCreatesTombstone() {
        val identity = identityFields()
        val empty = VaultPayload.empty("Test", "device-a").copy(revision = 1, updatedAt = "2026-01-01T00:00:00Z")
        val added = VaultMutations.addIdentity(empty, identity, "device-a", Instant.parse("2026-01-02T00:00:00Z"), "identity-one")
        assertEquals("Alex", (added.items.single() as IdentityItem).firstName)
        val edited = VaultMutations.editIdentity(added, "identity-one", identity.copy(company = "Updated"), "device-b", Instant.parse("2026-01-03T00:00:00Z"))
        assertEquals(2, (edited.items.single() as IdentityItem).revision)
        val favorite = VaultMutations.toggleIdentityFavorite(edited, "identity-one", "device-b", Instant.parse("2026-01-04T00:00:00Z"))
        assertTrue((favorite.items.single() as IdentityItem).favorite)
        val deleted = VaultMutations.deleteIdentity(favorite, "identity-one", "device-b", Instant.parse("2026-01-05T00:00:00Z"))
        assertTrue(deleted.items.isEmpty())
        assertEquals(Tombstone("identity-one", "2026-01-05T00:00:00Z", 5, "device-b"), deleted.tombstones.single())
    }

    @Test
    fun identityValidationAndDuplicateDetectionFailClosed() {
        val identity = identityFields()
        val empty = VaultPayload.empty("Test", "device-a")
        assertTrue(runCatching { VaultMutations.addIdentity(empty, identity.copy(firstName = "", middleName = "", lastName = "", email = "", phone = "", company = "", addressLine1 = "", addressLine2 = "", city = "", region = "", postalCode = "", country = ""), "device-a") }.isFailure)
        assertTrue(runCatching { VaultMutations.addIdentity(empty, identity.copy(email = "invalid"), "device-a") }.isFailure)
        val added = VaultMutations.addIdentity(empty, identity, "device-a", itemId = "identity-one")
        assertEquals(1, VaultMutations.likelyIdentityDuplicates(added, identity.copy(title = "Other", email = " ALEX@example.com ")).size)
    }

    private fun identityFields() = IdentityFields(
        title = "Personal identity", firstName = "Alex", middleName = "Q", lastName = "Person",
        email = "alex@example.com", phone = "+1 555 0100", company = "Example", addressLine1 = "1 Main Street",
        addressLine2 = "Unit 2", city = "Seattle", region = "WA", postalCode = "98101", country = "US", notes = "Primary",
    )
}

File: shared/src/vault-mutations.test.ts
import { describe, expect, it } from "vitest";
import { createEmptyVault } from "./models.js";
import { addCreditCard, addIdentity, addLogin, addSecureNote, deleteCreditCard, deleteIdentity, deleteLogin, deleteSecureNote, editCreditCard, editIdentity, editLogin, editSecureNote, findLikelyCreditCardDuplicates, findLikelyIdentityDuplicates, findLikelyLoginDuplicates, findLikelySecureNoteDuplicates, toggleCreditCardFavorite, toggleIdentityFavorite, toggleLoginFavorite, toggleSecureNoteFavorite, updateSecuritySettings } from "./vault-mutations.js";

const fields = {
  title: "Example",
  username: "person@example.com",
  password: "secret",
  uris: ["https://example.com/login"],
  androidPackageNames: ["com.example.app"],
};

describe("login mutations", () => {
  it("adds, edits, favorites, and tombstones a login with monotonic revisions and timestamps", () => {
    const empty = createEmptyVault("Test", "device-a", new Date("2026-01-01T00:00:00.000Z"));
    const added = addLogin(empty, fields, { deviceId: "device-a", now: new Date("2026-01-02T00:00:00.000Z"), itemId: "login-one" });
    expect(added).toMatchObject({ revision: 2, updatedAt: "2026-01-02T00:00:00.000Z", writerDeviceId: "device-a" });
    expect(added.items[0]).toMatchObject({ id: "login-one", revision: 1, createdAt: "2026-01-02T00:00:00.000Z" });

    const edited = editLogin(added, "login-one", { ...fields, title: "Updated" }, { deviceId: "device-b", now: new Date("2026-01-03T00:00:00.000Z") });
    expect(edited).toMatchObject({ revision: 3, writerDeviceId: "device-b" });
    expect(edited.items[0]).toMatchObject({ title: "Updated", revision: 2, createdAt: "2026-01-02T00:00:00.000Z", updatedAt: "2026-01-03T00:00:00.000Z" });

    const favorited = toggleLoginFavorite(edited, "login-one", { deviceId: "device-b", now: new Date("2026-01-04T00:00:00.000Z") });
    expect(favorited.items[0]).toMatchObject({ favorite: true, revision: 3 });

    const deleted = deleteLogin(favorited, "login-one", { deviceId: "device-b", now: new Date("2026-01-05T00:00:00.000Z") });
    expect(deleted.items).toHaveLength(0);
    expect(deleted.tombstones).toEqual([{ itemId: "login-one", deletedAt: "2026-01-05T00:00:00.000Z", revision: 5, deviceId: "device-b" }]);
  });

  it("warns for the same identifier on an exact origin or Android package", () => {
    const vault = addLogin(createEmptyVault("Test", "device-a"), fields, { deviceId: "device-a", itemId: "login-one" });
    expect(findLikelyLoginDuplicates(vault, { ...fields, uris: ["https://example.com/other"], androidPackageNames: [] })).toHaveLength(1);
    expect(findLikelyLoginDuplicates(vault, { ...fields, title: "Other", username: "other@example.com", uris: ["https://other.example"], androidPackageNames: [] })).toHaveLength(0);
  });

  it("persists security settings as one normal vault revision", () => {
    const empty = createEmptyVault("Test", "device-a", new Date("2026-01-01T00:00:00.000Z"));
    const updated = updateSecuritySettings(empty, 15, 60, { deviceId: "device-b", now: new Date("2026-01-02T00:00:00.000Z") });
    expect(updated).toMatchObject({
      revision: 2,
      updatedAt: "2026-01-02T00:00:00.000Z",
      writerDeviceId: "device-b",
      settings: { autoLockMinutes: 15, clearClipboardSeconds: 60 },
    });
    expect(() => updateSecuritySettings(updated, 0, 60, { deviceId: "device-b" })).toThrow(RangeError);
  });
});

describe("secure note mutations", () => {
  it("adds, edits, favorites, and tombstones a note without changing its identity", () => {
    const empty = createEmptyVault("Test", "device-a", new Date("2026-01-01T00:00:00.000Z"));
    const added = addSecureNote(empty, { title: "Recovery", body: "Offline codes" }, { deviceId: "device-a", now: new Date("2026-01-02T00:00:00.000Z"), itemId: "note-one" });
    expect(added.items[0]).toMatchObject({ id: "note-one", kind: "secureNote", revision: 1, title: "Recovery", body: "Offline codes" });

    const edited = editSecureNote(added, "note-one", { title: "Recovery codes", body: "Updated codes" }, { deviceId: "device-b", now: new Date("2026-01-03T00:00:00.000Z") });
    expect(edited.items[0]).toMatchObject({ id: "note-one", revision: 2, title: "Recovery codes", body: "Updated codes", createdAt: "2026-01-02T00:00:00.000Z" });

    const favorite = toggleSecureNoteFavorite(edited, "note-one", { deviceId: "device-b", now: new Date("2026-01-04T00:00:00.000Z") });
    expect(favorite.items[0]).toMatchObject({ favorite: true, revision: 3 });

    const deleted = deleteSecureNote(favorite, "note-one", { deviceId: "device-b", now: new Date("2026-01-05T00:00:00.000Z") });
    expect(deleted.items).toHaveLength(0);
    expect(deleted.tombstones).toContainEqual({ itemId: "note-one", deletedAt: "2026-01-05T00:00:00.000Z", revision: 5, deviceId: "device-b" });
  });

  it("validates content and detects duplicate titles", () => {
    const empty = createEmptyVault("Test", "device-a");
    expect(() => addSecureNote(empty, { title: "", body: "body" }, { deviceId: "device-a" })).toThrow(RangeError);
    expect(() => addSecureNote(empty, { title: "Title", body: "  " }, { deviceId: "device-a" })).toThrow(RangeError);
    const added = addSecureNote(empty, { title: "Recovery", body: "codes" }, { deviceId: "device-a", itemId: "note-one" });
    expect(findLikelySecureNoteDuplicates(added, { title: " recovery ", body: "different" })).toHaveLength(1);
    expect(findLikelySecureNoteDuplicates(added, { title: "Other", body: "different" })).toHaveLength(0);
  });
});

describe("credit card mutations", () => {
  const card = {
    title: "Travel card",
    cardholderName: "A Person",
    number: "4111 1111 1111 1111",
    expiryMonth: 12,
    expiryYear: 2030,
    verificationCode: "123",
    pin: "9876",
    notes: "Use abroad",
  };

  it("normalizes, edits, favorites, and tombstones a card", () => {
    const empty = createEmptyVault("Test", "device-a", new Date("2026-01-01T00:00:00.000Z"));
    const added = addCreditCard(empty, card, { deviceId: "device-a", now: new Date("2026-01-02T00:00:00.000Z"), itemId: "card-one" });
    expect(added.items[0]).toMatchObject({ id: "card-one", kind: "creditCard", number: "4111111111111111", revision: 1 });
    const edited = editCreditCard(added, "card-one", { ...card, title: "Primary card", notes: "Updated" }, { deviceId: "device-b", now: new Date("2026-01-03T00:00:00.000Z") });
    expect(edited.items[0]).toMatchObject({ title: "Primary card", notes: "Updated", revision: 2, createdAt: "2026-01-02T00:00:00.000Z" });
    const favorite = toggleCreditCardFavorite(edited, "card-one", { deviceId: "device-b", now: new Date("2026-01-04T00:00:00.000Z") });
    expect(favorite.items[0]).toMatchObject({ favorite: true, revision: 3 });
    const deleted = deleteCreditCard(favorite, "card-one", { deviceId: "device-b", now: new Date("2026-01-05T00:00:00.000Z") });
    expect(deleted.items).toHaveLength(0);
    expect(deleted.tombstones).toContainEqual({ itemId: "card-one", deletedAt: "2026-01-05T00:00:00.000Z", revision: 5, deviceId: "device-b" });
  });

  it("rejects invalid fields and detects the same normalized card number", () => {
    const empty = createEmptyVault("Test", "device-a");
    expect(() => addCreditCard(empty, { ...card, number: "123" }, { deviceId: "device-a" })).toThrow(RangeError);
    expect(() => addCreditCard(empty, { ...card, expiryMonth: 13 }, { deviceId: "device-a" })).toThrow(RangeError);
    expect(() => addCreditCard(empty, { ...card, verificationCode: "x" }, { deviceId: "device-a" })).toThrow(RangeError);
    const added = addCreditCard(empty, card, { deviceId: "device-a", itemId: "card-one" });
    expect(findLikelyCreditCardDuplicates(added, { ...card, number: "4111-1111-1111-1111" })).toHaveLength(1);
  });
});

describe("identity mutations", () => {
  const identity = {
    title: "Personal identity",
    firstName: "Alex",
    middleName: "Q",
    lastName: "Person",
    email: "alex@example.com",
    phone: "+1 555 0100",
    company: "Example",
    addressLine1: "1 Main Street",
    addressLine2: "Unit 2",
    city: "Seattle",
    region: "WA",
    postalCode: "98101",
    country: "US",
    notes: "Primary",
  };

  it("adds, edits, favorites, and tombstones an identity", () => {
    const empty = createEmptyVault("Test", "device-a", new Date("2026-01-01T00:00:00.000Z"));
    const added = addIdentity(empty, identity, { deviceId: "device-a", now: new Date("2026-01-02T00:00:00.000Z"), itemId: "identity-one" });
    expect(added.items[0]).toMatchObject({ id: "identity-one", kind: "identity", firstName: "Alex", revision: 1 });
    const edited = editIdentity(added, "identity-one", { ...identity, company: "Updated" }, { deviceId: "device-b", now: new Date("2026-01-03T00:00:00.000Z") });
    expect(edited.items[0]).toMatchObject({ company: "Updated", revision: 2, createdAt: "2026-01-02T00:00:00.000Z" });
    const favorite = toggleIdentityFavorite(edited, "identity-one", { deviceId: "device-b", now: new Date("2026-01-04T00:00:00.000Z") });
    expect(favorite.items[0]).toMatchObject({ favorite: true, revision: 3 });
    const deleted = deleteIdentity(favorite, "identity-one", { deviceId: "device-b", now: new Date("2026-01-05T00:00:00.000Z") });
    expect(deleted.items).toHaveLength(0);
    expect(deleted.tombstones).toContainEqual({ itemId: "identity-one", deletedAt: "2026-01-05T00:00:00.000Z", revision: 5, deviceId: "device-b" });
  });

  it("requires useful details and detects duplicate email", () => {
    const empty = createEmptyVault("Test", "device-a");
    const blank = Object.fromEntries(Object.keys(identity).map((key) => [key, key === "title" ? "Empty" : ""])) as typeof identity;
    expect(() => addIdentity(empty, blank, { deviceId: "device-a" })).toThrow(RangeError);
    expect(() => addIdentity(empty, { ...identity, email: "invalid" }, { deviceId: "device-a" })).toThrow(RangeError);
    const added = addIdentity(empty, identity, { deviceId: "device-a", itemId: "identity-one" });
    expect(findLikelyIdentityDuplicates(added, { ...identity, title: "Other", email: " ALEX@example.com " })).toHaveLength(1);
  });
});
